### ISSUE TITLE
Failed to find the GHC version of this Stack project. Error when calling stack setup --silent

### ISSUE THREAD

**Issue by anentropic** [haskell/vscode-haskell#841]
### Your environment

Which OS do you use: macOS

### Steps to reproduce

I have installed Haskell via `ghcup` and set the active version `ghcup set 9.6.1`

I have installed all the bits that ghcup said needed updating, including HLS.

I cloned the following repo https://github.com/BioHaskell/octree and opened it as a project in VS Code.

In the process of trying to get it working I have added the following config in `settings.json`:
```json
    "haskell.manageHLS": "GHCup",
    "haskell.toolchain": {
        "ghc": "9.6.1",
        "hls": "latest",
        "cabal": "recommended"
    },
```
but it didn't make any noticeable difference before/after

### Expected behaviour

No error, the extension does something useful.

### Actual behaviour

When I open VS Code into the project I get this error message:

```
Failed to find the GHC version of this Stack project. Error when calling stack setup --silent
```

and there don't seem to be any intellisense features working.

I have no idea how I am supposed to configure the extension for it to work, looking for guidance.

### Include debug information

Execute in the root of your project the command `haskell-language-server-wrapper --debug .` and paste the logs here (you can find the executable location [here](https://github.com/haskell/vscode-haskell#downloaded-binaries)):

<details>
<summary>
Debug output:
</summary>

```
haskell-language-server-wrapper --debug .
No 'hie.yaml' found. Try to discover the project type!
Run entered for haskell-language-server-wrapper(haskell-language-server-wrapper) Version 1.9.0.0 aarch64 ghc-9.4.4
Current directory: /Users/anentropic/Documents/Dev/Sources/octree
Operating system: darwin
Arguments: ["--debug","."]
Cradle directory: /Users/anentropic/Documents/Dev/Sources/octree
Cradle type: Stack

Tool versions found on the $PATH
cabal:          3.10.1.0
stack:          2.9.3
ghc:            9.6.1


Consulting the cradle to get project GHC version...
Failed to find the GHC version of this Stack project.
Error when calling stack setup --silent

[ Error ] [GHCup-00130] The version 8.6.5 of the tool ghc is not installed.
[ Info  ] downloading: https://raw.githubusercontent.com/haskell/ghcup-metadata/master/ghcup-0.0.7.yaml as file /Users/anentropic/.ghcup/cache/ghcup-0.0.7.yaml
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
[ Error ] [GHCup-00010] Unable to find a download for the requested version/distro.

 2023-03-27T21:42:35.011135Z | Info | Starting server

```

</details>

Paste the contents of extension specific log, you can check instructions about how to find it [here](https://github.com/haskell/vscode-haskell#troubleshooting)

<details>
<summary>
Extension log:
</summary>

Did not find any details of how to get the extension log at the link given

</details>

**Comment by hasufell** [comment]
It seems you're on Mac M1.

The project you're trying to build has an ancient stack resolver: `resolver: lts-14.4`. That one seems to correspond to GHC-8.6.5. Stack will try to install that version (that's how stack works). But M1 is only supported starting with GHC 8.10.x.

**Comment by anentropic** [comment]
do I have to specify a specific resolver? can I remove that line, or give it a value which means "use default", or "latest"?

**Comment by hasufell** [comment]
Check the stack docs: https://docs.haskellstack.org/en/stable/stack_yaml_vs_cabal_package_file/#resolvers-and-snapshots

**Comment by anentropic** [comment]
Is there no step-by-step guide anywhere?

How do I choose an appropriate resolver?

I tried deleting the `stack.yaml.lock` file which had been created, and commenting out that line in `stack.yaml` but I get same error

I tried `resolver: ghc-9.6.1` since that is what I'd installed via ghcup, but then I get the error:

```
Failed to find a HLS version for GHC 9.6.1
Executable names we failed to find: haskell-language-server-9.6.1,haskell-language-server
```

Scrolling back through my terminal history I found:
```
ghcup set 9.6.1
[ Warn  ] This is an old-style command for setting GHC. Use 'ghcup set ghc' instead.
[ Warn  ] GHC-9.6.1 appears to have no corresponding HLS-1.9.0.0 binary.
[ ...   ] Haskell IDE support may not work.
[ ...   ] You can try to either:
[ ...   ]   1. Install a different HLS version (e.g. downgrade for older GHCs)
[ ...   ]   2. Install and set one of the following GHCs: [9.4.4, 9.4.3, 9.2.5, 9.0.2, 8.10.7]
[ ...   ]   3. Let GHCup compile HLS for you, e.g. run: ghcup compile hls -g 1.9.0.0 --ghc 9.6.1 --cabal-update
[ ...   ]      (see https://www.haskell.org/ghcup/guide/#hls for more information)
[ Info  ] GHC 9.6.1 successfully set as default version
```

So I did:
```
$ ghcup set ghc
[ Info  ] GHC 9.2.5 successfully set as default version
```

I changed the stack.yaml to `resolver: ghc-9.5.2` and did a `cabal build` and then `code .` to reopen VS Code

This time no error 👍 

At first I thought there were no intellisense features either, but if I scroll to the top of one of the source files I can see a red squiggle with this error on hover:
```
<command line>: cannot satisfy -package linear
    (use -v for more information)compiler
```

I am wondering if I would be better off deleting the `stack.yaml` completely - would it then treat the `.cabal` file as authoritative?

**Comment by anentropic** [comment]
> I am wondering if I would be better off deleting the `stack.yaml` completely - would it then treat the `.cabal` file as authoritative?

I tried that...

if I reopen VS Code I get this again:
```
Failed to find a HLS version for GHC 9.6.1 Executable names we failed to find: haskell-language-server-9.6.1,haskell-language-server
```

So I remembered the config I'd added in `settings.json`, deleted the `haskell.toolchain` part

Reopened VS Code, the red squiggle now shows a new error:
```
Failed to run ["cabal","v2-repl","Octree:bench:bench-octree"] in directory "/Users/anentropic/Documents/Dev/Sources/octree". Consult the logs for full command and error.
Failed command: cabal --builddir=/Users/anentropic/.cache/hie-bios/dist-octree-51878d0c5bfcbf44451e6caa67f9c287 v2-repl --with-compiler /Users/anentropic/.cache/hie-bios/wrapper-b54f81dea4c0e6d1626911c526bc4e36 --with-hc-pkg /Users/anentropic/.cache/hie-bios/ghc-pkg-6ae5de20ada26caebd3ab08b3b228b91 Octree:bench:bench-octree

Error: cabal: Cannot open a repl for the benchmark 'bench-octree' because the
solver did not find a plan that included the benchmarks for Octree-0.6.1.0. It
is probably worth trying again with benchmarks explicitly enabled in the
configuration in the cabal.project{.local} file. This will ask the solver to
find a plan with the benchmarks available. It will either fail with an
explanation or find a different plan that uses different versions of some
other packages. Use the '--dry-run' flag to see package versions and check
that you are happy with the choices.
```

Presumably something to do with this section in the cabal file:
```cabal
benchmark bench-octree
  type:             exitcode-stdio-1.0
  other-modules:    Data.Octree.Internal
  main-is:          tests/benchmark.hs
  build-depends:    base >=4.0 && <4.13, deepseq, ghc-prim,
                    linear >= 1.18.0, QuickCheck >= 2.8.0,
                    lens   >= 4.13.0,
                    criterion
  extensions:       ScopedTypeVariables
```

It's probably the `base >=4.0 && <4.13` ...I already had to remove the upper bound from the `Library` target to get `cabal build` to succeed, seemed like `base < 4.13` is too old for compatibility with ghc 9.x, and I can't install ghc 8.x via ghcup because that's too old instead

I removed the `&& <4.13` from `base` for both the `Test-suite` and `benchmark` targets, and reopened VS Code again

No errors, and after a bit of a wait for packages to install in the background the intellisense is now working 👍
